### DISCUSSION TITLE
Session in multi cluster architecture

### DISCUSSION THREAD

**Discussion by dzagagnoni89** [phax/phoss-smp#232] [Q&A]
Hello,
we installed phoss-smp in Production where we have 2 replicated K8S clusters (active/active). In each cluster we have 2 replicas of the pod where the container runs with the webapp. 
We are having problems accessing the "secure" part as if we log into a pod, when we try to access another section of the webapp we are automatically logged out and have to enter the credentials again in a loop. 
Is there a way to handle the session not in memory? 
Thanks a lot in advance for your support!
Regards,
Davide

**Comment by phax** [comment]
Hi Davide,
so first of all, you have to use the SMP with SQL or MongoDB backend - XML backend doesn't work with horizontal scaling (because it loads the data in memory).
Potentially what you need to do, is to bind one HTTP Session to one Cluster Node. Evtl. the description at https://tomcat.apache.org/tomcat-9.0-doc/cluster-howto.html helps you.
If you want persistent sessions (never tried that) please refer to the respective Tomcat site: https://tomcat.apache.org/tomcat-9.0-doc/config/manager.html

hth, Philip

**Reply by dzagagnoni89** [reply]
Thanks a lot Philip,
yes we are using MongoDB Backend.
As for the solutions you have suggested, we will try the second way and I will let you know if it will solve our problem.
Thanks so much for the feedback!
Se you soon.
Davide

**Reply by phax** [reply]
Good luck - I hope Tomcat supports MongoDB as a persistence backend :)

**Comment by mySupply-JEB** [comment]
Hello @phax (and Davide),

We're running into a nearly identical "issue" and are looking into the possibilities of resolving this in a reasonable manner. 
In such setups it would be possible to fiddle with the load balancers to create sticky sessions that direct the traffic from a client to the same exact SMP instance once client has initiated the session. This is not the most fault-tolerant setup, though, as whenever a server running an SMP instance is shutdown, the session would be dropped. This becomes a particular problem for people and organizations who get their hosting capabilities on a public cloud spot-market (servers rapidly shutdown and start, get swapped like crazy, etc).

I think it would be most viable to achieve a high availability setup with the current implementation of the Phoss SMP by simply overriding the current manager configuration in the ${CATALINA_HOME]/conf/session.xml (which can be done in Docker/Kubernetes by creating a volume and mounting it on the same location, this should override the existing default configuration):

Example based on MySQL: 
```
<Manager className="org.apache.catalina.session.PersistentManager" 
         saveOnRestart="true">
  <Store className="org.apache.catalina.session.JDBCStore" 
         connectionURL="jdbc:mysql://<MYSQL_HOST>:<PORT>/<DATABASE>"
         driverName="com.mysql.cj.jdbc.Driver"
         connectionName="<DB_USER>"
         connectionPassword="<DB_PASSWORD>"
         sessionTable="tomcat_sessions"
         sessionIdCol="session_id"
         sessionValidCol="valid_session"
         sessionMaxInactiveCol="max_inactive"
         sessionLastAccessedCol="last_access"
         sessionDataCol="session_data"/>
</Manager>
```

And then we "just" need a migration to setup the table in the database. Should also be fairly simple, something like the following based on https://tomcat.apache.org/tomcat-10.1-doc/config/manager.html:
```
CREATE TABLE tomcat_sessions (
  session_id     VARCHAR(100) NOT NULL PRIMARY KEY,
  valid_session  CHAR(1) NOT NULL DEFAULT '1',
  max_inactive   INT NOT NULL,
  last_access    BIGINT NOT NULL,
  app_name       VARCHAR(255),
  session_data   MEDIUMBLOB,
  KEY idx_app_name(app_name)
);
```

Would you want to include such a migration, maybe "hidden" behind a boolean high-availability configuration? I am unable to make a proof-of-concept on the SMP at this very moment but would love to do so whenever I can as this has been a bit of a headache of mine for a while blocking us from getting a truly scalable SMP. I dislike the idea of _**not**_ having such a migration within the source application. :)

I have done very similar setups with .NET applications that have worked successfully and without issues.

// Jeppe Buthler, mySupply

**Reply by phax** [reply]
I created #424 out of it
